text/0000-style-fn-decl.md

```diff
@@ -0,0 +1,274 @@
+
+
+### Indenting return type
+
+Aligning the return type with the arguments makes it hard to scan for (one could
+argue that the argument type is just another component of the function type, but
+in practice I find myself often wanting to scan functions for the return type,
+or to quickly notice if a function returns something or not). Indenting the
+return type at block indent is hard for scanning if there is only one argument:
+
+```
+fn foo<A, B, C>(x: SomeType)
+-> ReturnType<A>
+{
+```
+
+Indenting at block indent + 4 works in all circumstances and is easy to scan for.
+
+
+### Where clauses
+
+The same arguments about indenting return types apply here. I believe
+distinguishing `->` from `where` is easy enough that having where clauses and
+return types at the same level of indent is OK.
+
+Where clause predicates can be long; aligning to block indent + 4 rather than to
+the arguments gives more space for long predicates.
+
+Starting the line with `where` makes the clause easy to scan for (as opposed by
+having it on the same line as the last argument).
+
+Aligning the type variables is tidy and consistent with arguments. Having one
+predicate per line, rather than sometimes putting the whole clause on one line
+is consistent and does not affect the principle of keeping small functions small,
+since a function with a multiple-predicate where clause is not a small function.
+Furthermore, predicates can be syntatically complex and separating them by line
+is useful for readability.
+
+
+### The opening brace
+
+Keeping the opening brace on the last line of the signature where there is a
+return type or where clause means that there is no whitespace between the
+function signature and body. E.g.,
+
+```
+fn foo<A, B, C>(x: SomeType)
+    -> ReturnType<A> {
+    let a = bar(x);
+    let b = baz(a, x);
+    b
+}
+```
+
+That is horrible for scanning (and just really ugly).
+
+For the sake of consistency, we should use the newline brace style for any
+multi-line function, even if there is no where clause or return type. I think it
+is OK to use same line brace for one line function signatures for the sake of
+keeping small functions small, but it seems that arguing for consistency here is
+also reasonable. E.g.,
+
+```
+fn foo<T>(x: T, y: A) -> T
+{
+```
+
+Note that both same-line and newline brace styles are common in most programming
+languages, although usually they are not mixed. Also note that the style we use
+for functions does not necessarily have to affect the style used for block
+expressions (e.g., the Mozilla C++ style guide requires newline braces for all
+functions and same-line braces for all block statements).
+
+
+## Questions
+
+* Should there be a terminating comma for where clauses?
+* Should the opening brace always be on a new line? What about one line signatures (the shortest examples)?
+* Should where clauses always get their own line, even for short functions (see alternatives in examples)?
```

Comment:
Yes, they blur into the return type too easily otherwise.

```rust
fn foo<T>(&mut self, input_a: Foo, deltarizer: Bar) -> Frob<T> where T: Clone + Ord {
```

Fundamentally any `where` that is simple enough to merit "on the same line" seems like it should fall under "shouldn't have been a where"